### PR TITLE
#1822 Add detailed data to constraints

### LINKED ISSUE #1822: Extend constraints to add structured data
Currently, detailed information about each `Constraint` is stored in its field `description` as text.

It is proposed to extend the `Constraint` class to add relevant data as explicit attributes. For limit violations: the type of limit, the value of the limit, the value of the magnitude. For current limit violations: the side of the equipment and the acceptable duration if it was a temporary limit.

Adding this information in explicit fields will allow an improved integration of Dynawo as a simulation tool for security analysis. Instead of relying on processing the final state output of Dynawo for each contingency, the clients of Dynawo as a security analysis engine could receive only the summary of limit violations that have been observed.

### PR DESCRIPTION
**Pull request by zamarrenolm**
Closes #1822

### CODE REVIEW

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp [hunk 1]
```diff
@@ -52,6 +52,11 @@ ModelCurrentLimits::setSide(const side_t side) {
   side_ = side;
 }
 
+void
+ModelCurrentLimits::setFactorPuToA(const double& factorPuToA) {
```
Comment (lecourtoisflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp:56]: do not use const& for native types

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp [hunk 2]
```diff
@@ -83,36 +89,58 @@ ModelCurrentLimits::evalG(const double& t, const double& current, state_g* g, co
   }
 }
 
+constraints::ConstraintData
+ModelCurrentLimits::constraintData(std::string kind, int i) {
```
Comment (lecourtoisflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp:93]: use unsigned int for index and const string& or enum for kind

**Comment by zamarrenolm** [comment]
Commits squashed and branch rebased on master

**Comment by zamarrenolm** [comment]
Changes rebased on master

File: dynawo/sources/API/CSTR/CSTRConstraint.h [hunk 3]
```diff
@@ -22,10 +22,24 @@
 
 #include "CSTRConstraintCommon.h"
 
+#include <boost/optional.hpp>
 #include <string>
 
 namespace constraints {
 
+struct ConstraintData {
+  std::string kind;
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/API/CSTR/CSTRConstraint.h:31]: Seeing what you do in powsybl-dynawo to handle the kind type, ie use an enum, I wonder if we should do it here also

**Review by gautierbureau — COMMENTED** [review]
other than the enum it seems ok for me

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp [hunk 4]
```diff
@@ -83,36 +89,58 @@ ModelCurrentLimits::evalG(const double& t, const double& current, state_g* g, co
...
       if (g[0 + 2 * i] == ROOT_UP && !activated_[i]) {
         if (openingAuthorized_[i]) {  // Delay is specified => temporary limit
-          DYNAddConstraint(network, componentName,  true,  modelType, OverloadUp, acceptableDurations_[i], side_);
+          DYNAddConstraintWithData(network, componentName, true, modelType,
+            constraintData("OverloadUp", i),
+            OverloadUp, acceptableDurations_[i], side_);
         } else {
-          DYNAddConstraint(network, componentName,  true,  modelType, IMAP, side_);
+          DYNAddConstraintWithData(network, componentName, true, modelType,
+            constraintData("IMAP", i),
```
Comment (marcochiaramello) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp:118]: IMAP is a french acronym, PATL is the english translation

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp [hunk 5]
```diff
@@ -83,36 +89,58 @@ ModelCurrentLimits::evalG(const double& t, const double& current, state_g* g, co
...
       if (g[0 + 2 * i] == ROOT_DOWN && activated_[i]) {
         if (openingAuthorized_[i]) {  // Delay is specified => temporary limit
-          DYNAddConstraint(network, componentName,  false,  modelType, OverloadUp, acceptableDurations_[i], side_);
+          DYNAddConstraintWithData(network, componentName, false, modelType,
+            constraintData("OverloadUp", i),
+            OverloadUp, acceptableDurations_[i], side_);
         } else {
-          DYNAddConstraint(network, componentName,  false,  modelType, IMAP, side_);
+          DYNAddConstraintWithData(network, componentName, false, modelType,
+            constraintData("IMAP", i),
```
Comment (marcochiaramello) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp:132]: please, change IMAP to PATL, everywhere else

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.h [hunk 6]
```diff
@@ -114,10 +122,20 @@ class ModelCurrentLimits {  ///< Generic Current Limits model
...
+   * @return details for the constraint
+   */
+  constraints::ConstraintData constraintData(const std::string& kind, unsigned int i);
+
   int nbTemporaryLimits_;  ///< number of temporary limits (limits with a time duration)
   side_t side_;  ///< side
 
   double maxTimeOperation_;  ///< maximum time operation, if limits duration is over this time, the current limit does not operate
+  double lastCurrentValue_;  ///< last value of the current, kept to be reported in constraints
+  double factorPuToA_;  ///< factor to convert pu values to Amperes
```
Comment (marcochiaramello) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.h:138]: replace pu by p.u. or per unit

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.h (hunk 6 above)
Comment (zamarrenolm) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.h]: Replaced "pu" with "p.u."

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp (hunk 4 above)
Comment (zamarrenolm) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp]: Renamed IMAP to PATL (Permanent Admissible Transmission Loading)

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp (hunk 5 above)
Comment (zamarrenolm) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelCurrentLimits.cpp]: Renamed everywhere

**Comment by zamarrenolm** [comment]
> other than the enum it seems ok for me

I have changed the type for `kind` attribute in constraint detailed data from string to enum.

**Comment by gautierbureau** [comment]
> > other than the enum it seems ok for me
> 
> I have changed the type for `kind` attribute in constraint detailed data from string to enum.

Thanks. Could you squash ?

File: dynawo/sources/API/CSTR/CSTRConstraint.h [hunk 7]
```diff
@@ -22,10 +22,33 @@
 
 #include "CSTRConstraintCommon.h"
 
+#include <boost/optional.hpp>
 #include <string>
 
 namespace constraints {
 
+struct ConstraintData {
+  ///< enum of possible values for kind
```
Comment (lecourtoisflo) [inline dynawo/sources/API/CSTR/CSTRConstraint.h:31]: Comment is incorrect: remove '<'

**Review by lecourtoisflo — APPROVED** [review]

**Review by marcochiaramello — APPROVED** [review]

File: dynawo/sources/API/CSTR/CSTRConstraint.h (hunk 7 above)
Comment (zamarrenolm) [reply dynawo/sources/API/CSTR/CSTRConstraint.h]: fixed the comment

**Comment by zamarrenolm** [comment]
changes squashed and rebased on master

**Review by gautierbureau — APPROVED** [review]

**Pull request merged**